fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -43,33 +43,53 @@ impl<T> Interpreter<PredicateStorage, T> {
     ///
     /// This is not a valid entrypoint for debug calls. It will only return a `bool`, and not the
     /// VM state required to trace the execution steps.
-    pub fn check_predicates<Tx>(checked: Checked<Tx>, params: ConsensusParameters, gas_costs: GasCosts) -> bool
+    pub fn check_predicates<Tx>(
+        checked: Checked<Tx>,
+        params: ConsensusParameters,
+        gas_costs: GasCosts,
+    ) -> Result<PredicatesChecked, PredicateVerificationFailed>
     where
-        Tx: ExecutableTransaction,
+        Tx: ExecutableTransaction + fuel_tx::field::GasLimit,
         <Tx as IntoChecked>::Metadata: CheckedMetadata,
     {
-        let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
-
         if !checked.transaction().check_predicate_owners() {
-            return false;
+            return Err(PredicateVerificationFailed::InvalidOwner);
         }
 
-        #[allow(clippy::needless_collect)]
+        let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
+
         // Needed for now because checked is only freed once the value is collected into a Vec
-        let predicates: Vec<RuntimePredicate> = checked
-            .transaction()
-            .inputs()
-            .iter()
-            .enumerate()
-            .filter_map(|(idx, _)| RuntimePredicate::from_tx(&params, checked.transaction(), idx))
+        #[allow(clippy::needless_collect)]
+        let predicates: Vec<_> = (0..checked.transaction().inputs().len())
+            .filter_map(|i| RuntimePredicate::from_tx(&params, checked.transaction(), i))
             .collect();
 
-        predicates
-            .into_iter()
-            .fold(vm.init_predicate(checked), |result, predicate| -> bool {
-                // VM is cloned because the state should be reset for every predicate verification
-                result && vm.clone()._check_predicate(predicate)
-            })
+        // Since we reuse the vm objects otherwise, we need to keep the actual gas here
+        let tx_gas_limit = *checked.transaction().gas_limit();
+        let mut remaining_gas = tx_gas_limit;
+
+        vm.init_predicate(checked);
+
+        for predicate in predicates {
+            // VM is cloned because the state should be reset for every predicate verification
+            let mut vm = vm.clone();
+
+            vm.context = Context::Predicate { program: predicate };
+            vm.registers[REG_GGAS] = remaining_gas;
```

Comment:
Could we add `set_remaining_gas` and `get_remaining_gas` methods and reuse them here(and in the `init` function too)? In this case, we can hide work with registers(and we will not expose unique information that we need to set 2 registers).

Reply:
We would need two getters, since remaining gas could mean either GGAS or CGAS. I don't really see any point in abstracting this away.

Reply:
`set_remaining_gas` should be useful=)

Regarding `remaining_gas`, seems we already have this kind of the function 
<img width="599" alt="image" src="https://user-images.githubusercontent.com/18346821/213551969-6b1cc2d2-1c91-48da-984e-141c5a02a4f7.png">

Reply:
I disagree about usefulness, since the resulting function is quite error prone; it's rather rare to set both CGAS and GGAS. Added the function anyways: 0669c3358c4722b55b00d30d2780424d62b0788a